internal/datastore/crdb/pool/balancer.go

```diff
@@ -0,0 +1,214 @@
+}
+
+// nodeConnectionBalancer is generic over underlying connection types for
+// testing purposes. Callers should use the exported NodeConnectionBalancer
+type nodeConnectionBalancer[P balancePoolConn[C], C balanceConn] struct {
+	ticker        *time.Ticker
+	pool          balanceablePool[P, C]
+	healthTracker *NodeHealthTracker
+	rnd           *rand.Rand
+	seed          int64
+}
+
+// newNodeConnectionBalancer is generic over underlying connection types for
+// testing purposes. Callers should use the exported NewNodeConnectionBalancer.
+func newNodeConnectionBalancer[P balancePoolConn[C], C balanceConn](pool balanceablePool[P, C], healthTracker *NodeHealthTracker, interval time.Duration) *nodeConnectionBalancer[P, C] {
+	seed := int64(new(maphash.Hash).Sum64())
+	return &nodeConnectionBalancer[P, C]{
+		ticker:        time.NewTicker(interval),
+		healthTracker: healthTracker,
+		pool:          pool,
+		seed:          seed,
+		rnd:           rand.New(rand.NewSource(seed)),
+	}
+}
+
+// Prune starts periodically checking idle connections and killing ones that are determined to be unbalanced.
+func (p *nodeConnectionBalancer[P, C]) Prune(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			p.ticker.Stop()
+			return
+		case <-p.ticker.C:
+			p.pruneConnections(ctx)
+			runtime.GC()
+		}
+	}
+}
+
+// pruneConnections prunes connections to nodes that have more than MaxConns/(# of nodes)
+// This causes the pool to reconnect, which over time will lead to a balanced number of connections
+// across each node.
+func (p *nodeConnectionBalancer[P, C]) pruneConnections(ctx context.Context) {
+	conns := p.pool.AcquireAllIdle(ctx)
+	defer func() {
+		// release all acquired idle conns back
+		for _, c := range conns {
+			c.Release()
+		}
+	}()
+
+	// bucket connections by healthy node
+	healthyConns := make(map[uint32][]P, 0)
+	for _, c := range conns {
+		id := p.pool.Node(c.Conn())
+		if healthyConns[id] == nil {
+			healthyConns[id] = make([]P, 0, 1)
+		}
+		healthyConns[id] = append(healthyConns[id], c)
+	}
+
+	nodeCount := uint32(p.healthTracker.HealthyNodeCount())
+	if nodeCount == 0 {
+		nodeCount = 1
+	}
+
+	connectionCounts := make(map[uint32]uint32)
+	p.pool.Range(func(conn C, nodeID uint32) {
+		connectionCounts[nodeID]++
+	})
+
+	log.Ctx(ctx).Trace().
+		Str("pool", p.pool.ID()).
+		Any("counts", connectionCounts).
+		Msg("connections per node")
+
+	// Delete metrics for nodes we no longer have connections for
+	p.healthTracker.Lock()
+	for node := range p.healthTracker.nodesEverSeen {
+		// TODO: does this handle network interruptions correctly?
+		if _, ok := connectionCounts[node]; !ok {
+			connectionsPerCRDBNodeCountGauge.DeletePartialMatch(map[string]string{
+				"pool":    p.pool.ID(),
+				"node_id": strconv.FormatUint(uint64(node), 10),
+			})
+		}
+	}
+	p.healthTracker.Unlock()
+
+	nodes := maps.Keys(connectionCounts)
+	slices.Sort(nodes)
```

Comment:
do you need to run this sort if you are about to shuffle things around, or is this needed for testing purposes

Reply:
I need the shuffle to start from the same sorted set every time in order to be deterministic

Reply:
I could memoize the sort+shuffle call too